### ISSUE TITLE
Make it easier to rename files while downloading

### ISSUE THREAD

**Issue by sschuberth** [michel-kraemer/gradle-download-task#245]
**Describe the solution you'd like**

When downloading multiple files specified via the `src` property, it would be nice if the `dest` property could also point to a collection of the same size containing the target file names. Currently, `dest` has to be a single value denoting the download directory.

**Describe alternatives you've considered**

I've read the [recipes](https://michelkraemer.com/recipes-for-gradle-download/#recipe-3-download-and-rename-multiple-files), but both alternatives don't look very elegant to me, esp. as they seem to prevent parallel downloads.

**Comment by michel-kraemer** [comment]
I'm not sure there's an easy way to change the interface without causing incompatibility issues. I'm currently thinking about alternatives. One idea that came to my mind is to add an `eachFile` action similar to the `Copy` task (https://docs.gradle.org/current/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:eachFile(org.gradle.api.Action)) but I think this needs to be carefully thought through and, once implemented, tested thoroughly.

**Comment by michel-kraemer** [comment]
The second option from the recipes does not prevent parallel downloads by the way.

**Comment by sschuberth** [comment]
> The second option from the recipes does not prevent parallel downloads by the way.

I'm, however, having some problems leveraging that approach, as the list of URL is generated from a JSON that I need to download first in my case. So I can't "statically" create those per-URL download tasks as the list of URLs is not known yet, but I also cannot dynamically create those tasks from the "parent" tasks that downloads and parses the JSON, as Gradle does not seem to support creating tasks from within other tasks (I'm getting `DefaultTaskContainer#register(String, Class, Action) on task set cannot be executed in the current context`).

Any recipe idea for that use-case?

**Comment by sschuberth** [comment]
> Any recipe idea for that use-case?

I'm now actually doing this via a `finalizedBy` task that [renames the files](https://github.com/oss-review-toolkit/ort/pull/5770/commits/44d6d560c540441a2b324eb492b3ffc48abde137#diff-2b2d0e698d405fc5b02de270fa4555d7f0c1692b6a2f601eb50776754802cc26R298-R299) in bulk after the (parallel) download.

**Comment by michel-kraemer** [comment]
Very cool idea!

I've added a `runAsync` method to the download extension, so you could basically do something like this and download multiple files in parallel:

```gradle
task downloadMultipleFiles1 {
    doLast {
        def fs = src.collect { s ->
            download.runAsync {
                src s.key
                dest new File("$buildDir/alternative1", s.value)
            }
        }
        CompletableFuture.allOf(fs as CompletableFuture[]).join()
    }
}
```

I haven't released a new version yet. I'd like to put some thought into this `eachFile` idea first.

**Comment by michel-kraemer** [comment]
How would you like this?

```gradle
ext.urls = [
    'http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.guess;hb=HEAD': 'config.guess',
    'http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.sub;hb=HEAD': 'config.sub'
]

task downloadMultipleFiles(type: Download) {
    src urls.keySet()
    dest buildDir
    eachFile { details ->
        details.name = urls[details.sourceURL.toString()]
    }
}
```

**Comment by michel-kraemer** [comment]
I've just released version 5.2.0 containing the new `eachFile` feature.
